File: js/src/detector.js
import { existsSync, statSync } from 'fs';
import { join, dirname } from 'path';
import { runCommand } from './commandRunner.js';

// Maps a VCS root directory to its system name.
// Populated on first detection; avoids repeated directory walks for the same repo.
const _rootCache = new Map();

/**
 * Directories whose answer needed the Perforce probe: the ones with no marker
 * directory anywhere above them.
 *
 * The root cache above only ever learns the positive answers, so a path outside
 * any working copy was re-walked AND re-probed on every call - and the probe is
 * a `p4 info` SPAWN, because Perforce has no marker directory to look for. A
 * tool that autosaves paid two of those per write (`writeTextFile` asks for the
 * provider twice), which measured at ~22ms per save on macOS. Nothing about
 * that answer changes between two keystrokes.
 *
 * Keyed by the EXACT starting directory and never consulted for ancestors,
 * unlike `_rootCache`. That distinction matters: "no marker above this
 * directory" says nothing about a repository nested somewhere below it, and an
 * ancestor walk over these would shadow one.
 */
const _probedCache = new Map();

/** Clear the detection cache. Called when the provider override is cleared. */
export function clearDetectorCache() {
  _rootCache.clear();
  _probedCache.clear();
}

/**
 * Auto-detect the active version control system for a given file path.
 * Walks up the directory tree looking for VC markers, then checks for Perforce.
 * Falls back to 'filesystem' if nothing is found.
 *
 * @param {string} absPath - Absolute path of the file (may not exist yet).
 * @returns {'git' | 'plastic' | 'svn' | 'perforce' | 'filesystem'}
 */
export function detectProvider(absPath) {
  let dir;
  try {
    dir = statSync(absPath).isDirectory() ? absPath : dirname(absPath);
  } catch {
    // Path doesn't exist yet — use its parent directory.
    dir = dirname(absPath);
  }

  const known = _probedCache.get(dir);
  if (known !== undefined) return known;

  // Check whether any ancestor is a known VCS root before doing any I/O.
  let current = dir;
  while (true) {
    if (_rootCache.has(current)) return _rootCache.get(current);
    const parent = dirname(current);
    if (parent === current) break;
    current = parent;
  }

  // Walk up looking for VC marker directories/files.
  current = dir;
  while (true) {
    if (existsSync(join(current, '.git'))) {
      _rootCache.set(current, 'git');
      return 'git';
    }
    if (existsSync(join(current, '.plastic'))) {
      _rootCache.set(current, 'plastic');
      return 'plastic';
    }
    if (existsSync(join(current, '.svn'))) {
      _rootCache.set(current, 'svn');
      return 'svn';
    }
    const parent = dirname(current);
    if (parent === current) break;
    current = parent;
  }

  // Perforce has no marker directory; detect by running `p4 info`.
  const p4 = runCommand('p4', ['info'], { timeout: 3000 });
  const answer = (p4.exitCode === 0 && p4.output.includes('Client name:')) ? 'perforce' : 'filesystem';
  // Remember it against this directory: a marker cannot appear in an ancestor
  // without somebody creating a working copy around a project that is already
  // open, and that is a restart either way.
  _probedCache.set(dir, answer);
  return answer;
}

File: csharp/SimpleVCLib/Detector.cs
using System.Text.Json;

namespace SimpleVCLib;

/// <summary>
/// Auto-detects the active version control system for a given path.
/// </summary>
public static class Detector
{
    private static readonly string[] ValidSystems =
        ["git", "perforce", "plastic", "svn", "filesystem"];

    // Maps a VCS root directory to its provider instance.
    // Populated on first detection; avoids repeated directory walks for the same repo.
    private static readonly Dictionary<string, IVCProvider> _rootCache = new();

    /// <summary>
    /// Directories whose answer needed the Perforce probe: the ones with no
    /// marker directory anywhere above them.
    /// <para>
    /// The root cache above only ever learns the POSITIVE answers, so a path
    /// outside any working copy was re-walked and re-probed on every call, and
    /// the probe is a `p4 info` subprocess. A tool that autosaves paid one per
    /// write. Nothing about that answer changes between two keystrokes.
    /// </para>
    /// <para>
    /// Keyed by the EXACT starting directory and never consulted for ancestors,
    /// unlike the root cache: "no marker above this directory" says nothing
    /// about a repository nested somewhere below it.
    /// </para>
    /// </summary>
    private static readonly Dictionary<string, IVCProvider> _probedCache = new();

    /// <summary>Clear the detection cache. Called when the provider override is cleared.</summary>
    public static void ClearCache()
    {
        _rootCache.Clear();
        _probedCache.Clear();
    }

    /// <summary>
    /// Detect and return the appropriate provider for <paramref name="path"/>.
    /// <para>
    /// Detection order:
    /// <list type="number">
    ///   <item>SIMPLE_VC environment variable</item>
    ///   <item>.vcconfig JSON file, walking up from the file's directory</item>
    ///   <item>VC marker directories (.git, .plastic, .svn)</item>
    ///   <item>Perforce (via `p4 info`)</item>
    ///   <item>Filesystem fallback</item>
    /// </list>
    /// </para>
    /// </summary>
    public static IVCProvider Detect(string path)
    {
        var startDir = GetDirectory(path);

        var envSystem = Environment.GetEnvironmentVariable("SIMPLE_VC")?.Trim().ToLowerInvariant();
        if (!string.IsNullOrEmpty(envSystem))
        {
            var p = CreateProvider(envSystem);
            if (p is not null) return p;
        }

        var dir = startDir;
        while (dir is not null)
        {
            var configPath = Path.Combine(dir, ".vcconfig");
            if (File.Exists(configPath))
            {
                var p = TryLoadConfig(configPath);
                if (p is not null) return p;
            }
            var parent = Path.GetDirectoryName(dir);
            if (parent == dir) break;
            dir = parent;
        }

        if (startDir is not null && _probedCache.TryGetValue(startDir, out var probed)) return probed;

        // Check whether any ancestor is a known VCS root before doing any I/O.
        dir = startDir;
        while (dir is not null)
        {
            if (_rootCache.TryGetValue(dir, out var cached)) return cached;
            var parent = Path.GetDirectoryName(dir);
            if (parent == dir) break;
            dir = parent;
        }

        // Walk up looking for VC marker directories/files.
        dir = startDir;
        while (dir is not null)
        {
            if (Directory.Exists(Path.Combine(dir, ".git")) ||
                File.Exists(Path.Combine(dir, ".git")))          // git worktrees use a file
            {
                var p = new GitProvider();
                _rootCache[dir] = p;
                return p;
            }

            if (Directory.Exists(Path.Combine(dir, ".plastic")))
            {
                var p = new PlasticProvider();
                _rootCache[dir] = p;
                return p;
            }

            if (Directory.Exists(Path.Combine(dir, ".svn")))
            {
                var p = new SvnProvider();
                _rootCache[dir] = p;
                return p;
            }

            var parent = Path.GetDirectoryName(dir);
            if (parent == dir) break;
            dir = parent;
        }

        // Perforce has no marker directory — detect via CLI.
        var p4 = CommandRunner.Run("p4", ["info"], timeoutMs: 3000);
        IVCProvider answer = (p4.ExitCode == 0 && p4.Output.Contains("Client name:"))
            ? new PerforceProvider()
            : new FilesystemProvider();
        if (startDir is not null) _probedCache[startDir] = answer;
        return answer;
    }

    // -------------------------------------------------------------------------
    // Helpers
    // -------------------------------------------------------------------------

    private static string GetDirectory(string path)
    {
        if (Directory.Exists(path)) return path;
        if (File.Exists(path))      return Path.GetDirectoryName(path) ?? path;
        // Path doesn't exist yet — treat as a file and use its parent directory.
        return Path.GetDirectoryName(path) ?? path;
    }

    private static IVCProvider? TryLoadConfig(string configPath)
    {
        try
        {
            var json = File.ReadAllText(configPath);
            using var doc = JsonDocument.Parse(json);
            if (doc.RootElement.TryGetProperty("system", out var systemElement))
            {
                var system = systemElement.GetString()?.Trim().ToLowerInvariant();
                if (!string.IsNullOrEmpty(system))
                    return CreateProvider(system);
            }
        }
        catch { /* Malformed config — ignore */ }
        return null;
    }

    private static IVCProvider? CreateProvider(string system) => system switch
    {
        "git"        => new GitProvider(),
        "perforce"   => new PerforceProvider(),
        "plastic"    => new PlasticProvider(),
        "svn"        => new SvnProvider(),
        "filesystem" => new FilesystemProvider(),
        _            => null,
    };
}
